src/sessions.ts

```diff
@@ -397,6 +457,49 @@ function isUnknownTransactionCommitResult(err: MongoError) {
   );
 }
 
+export function maybeClearPinnedConnection(
+  session: ClientSession,
+  options?: EndSessionOptions
+): void {
+  // unpin a connection if it has been pinned
+  const conn = session[kPinnedConnection];
+  const error = options?.error;
+
+  if (
+    session.inTransaction() &&
+    error &&
+    error instanceof MongoError &&
+    error.hasErrorLabel('TransientTransactionError')
+  ) {
+    return;
+  }
+
+  // NOTE: the spec talks about what to do on a network error only, but the tests seem to
+  //       to validate that we don't unpin on _all_ errors?
+  if (conn && (options?.error == null || options?.force)) {
+    const servers = Array.from(session.topology.s.servers.values());
+    if (servers.length === 0) {
+      // This can happen if the client is closed when the connection is still pinned
+      // NOTE: we don't usually do this, we could instead throw an error?
```

Comment:
Before we decide on whether to throw or just destory is there a snippet that can make this code reachable? Where would I close the client such that a connection is still pinned?

Reply:
This would cause this to get called in lb mode.

```
const cursor = collection.find();
cursor.next((err, doc) => {
  client.close();
});
```

Reply:
This code already throws with `MongoDriverError: Attempted to use a session that has ended` which makes sense because close is an async operation that can't stop .next() from moving on to the next document. I tried without awaiting close:

```js
for await (const doc of cursor) {
  client.close()
}
```
which throws 
```
./src/sdam/server.ts:330
      callback(new MongoDriverError('server is closed'));
```

And awaiting close
```js
for await (const doc of cursor) {
  await client.close()
}
```
which throws:
```
./src/cmap/connection_pool.ts:323
          waitQueueMember.callback(new MongoDriverError('connection pool closed'));
```

So I guess we have a precedent of throwing, I also don't know how safe it is to call destroy here without providing a callback. Don't we need to confirm when destroy is complete? (I'm actually not sure, it might be fine). If we do throw here we need maybClearPinnedConnection to take a callback or it could return the error? Or if we destroy then we still might need to involve a callback here, thoughts?

Reply:
It seems maybe more correct that we probably ensure destroy finishes, as it also removes the listeners from the pinned connection, and then we return a better worded error in the callback. Would you be ok with that?

Reply:
I took a look at the usages its going to be quite the refactor to make this function accept a callback, its doable but ClientSession#unpin for example doesn't currently take a callback so that will need it too, which branch out to its usages.

Throwing synchronously I think would also cause its own set of problems where `promise.catch()`s wouldn't work correctly.

The downside of not destroying is that we're leaking this connection right? If someone calls close while looping their cursor then .endSessions is called: https://github.com/mongodb/node-mongodb-native/blob/4.0/src/sdam/topology.ts#L486-L488 but it isn't currently awaited (or callback used) so even if we piped through the callback there would never be one defined 😕 this is tricky. 

I'm inclined we throw synchronously here and come back to fix this up? (@dariakp just want a second 👀 on this weird situation)

Reply:
I don't fully understand the implications of either side here... can connections jump between multiple pools? If not, it seems that there is no reason for a pool or the original server not to keep track of pinned connections so that when the server or the pool is destroyed, that connection is taken care of along with all the rest. I am assuming since the server length is 0, the actual associated server and pool was, in fact, destroyed by some existing mechanism.

Reply:
I guess I also want to understand 1) If we remove this code entirely, what's the undesirable result that may happen? and 2) If we leave this code as is and don't throw or call back with an error, what's the undesirable result that may happen? so, ultimately, the question is 3) In the ideal scenario, if we do reach this point, what's end state that we want to propagate to the user?

Reply:
I think the end state that we want is that 1) the connection is destroyed (and finishes that process so the event listeners are removed and the socket is closed, and 2) we give the user a meaningful error about what they did to get in that state.

So if we remove entirely, we are leaking connections as per point 1. If we leave as is, the user gets an error that might not truly explain why they got it. So I think the suggestion by @nbbeeken to just throw for now is the simplest solution in order to refactor into some callback craziness.

Reply:
> Just dropping a note from standup, I think the fact that we don't await endSessions in the topology.close method is a bug and causes this servers.length === 0 case to happen.

Reply:
My latest changes to address this (make sure all `endSession` calls finished before moving forward in the topology) didn't prove to be working, so I'll revert and make another ticket.

Reply:
This is now resolved as per the suggestion of waiting for all the end session calls to complete during closing of the topology.